Content.Server/Ninja/Systems/SpaceNinjaSystem.cs

```diff
@@ -0,0 +1,358 @@
+    [Dependency] private readonly SharedSubdermalImplantSystem _implants = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<SpaceNinjaComponent, ComponentStartup>(OnNinjaStartup);
+        SubscribeLocalEvent<SpaceNinjaComponent, MindAddedMessage>(OnNinjaMindAdded);
+        SubscribeLocalEvent<SpaceNinjaComponent, EmaggedSomethingEvent>(OnDoorjack);
+    }
+
+    public override void Update(float frameTime)
+    {
+        var query = EntityQueryEnumerator<SpaceNinjaComponent>();
+        while (query.MoveNext(out var uid, out var ninja))
+        {
+            UpdateNinja(uid, ninja, frameTime);
+        }
+    }
+
+    /// <summary>
+    /// Turns the player into a space ninja
+    /// </summary>
+    public void MakeNinja(Mind.Mind mind)
+    {
+        if (mind.OwnedEntity == null)
+            return;
+
+        // prevent double ninja'ing
+        var user = mind.OwnedEntity.Value;
+        if (HasComp<SpaceNinjaComponent>(user))
+            return;
+
+        AddComp<SpaceNinjaComponent>(user);
+        SetOutfitCommand.SetOutfit(user, "SpaceNinjaGear", EntityManager);
+        GreetNinja(mind);
+    }
+
+    /// <summary>
+    /// Download the given set of nodes, returning how many new nodes were downloaded.'
+    /// </summary>
+    public int Download(EntityUid uid, List<string> ids)
+    {
+        if (!GetNinjaRole(uid, out var role))
+            return 0;
+
+        var oldCount = role.DownloadedNodes.Count;
+        role.DownloadedNodes.UnionWith(ids);
+        var newCount = role.DownloadedNodes.Count;
+        return newCount - oldCount;
+    }
+
+    /// <summary>
+    /// Gets a ninja's role using the player's mind
+    /// </summary>
+    public bool GetNinjaRole(Mind.Mind? mind, [NotNullWhen(true)] out NinjaRole? role)
+    {
+        role = null;
+        if (mind == null)
+            return false;
+
+        foreach (var r in mind.AllRoles)
+        {
+            if (r is NinjaRole ninja)
+            {
+                role = ninja;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Gets a ninja's role using the player's entity id
+    /// </summary>
+    public bool GetNinjaRole(EntityUid uid, [NotNullWhen(true)] out NinjaRole? role)
+    {
+        role = null;
+        if (!TryComp<MindContainerComponent>(uid, out var mind))
+            return false;
+
+        return GetNinjaRole(mind.Mind, out role);
+    }
+
+    /// <summary>
+    /// Returns the global ninja gamerule config
+    /// </summary>
+    public NinjaRuleComponent RuleConfig()
+    {
+        var rule = EntityQuery<NinjaRuleComponent>().FirstOrDefault();
+        if (rule != null)
+            return rule;
+
+        _gameTicker.StartGameRule("Ninja", out var ruleEntity);
```

Comment:
Game rules shouldn't be treated as singletons, I think any existing examples are just holdovers from when gamerules actually were but now they should support being added as many times as we want.

Reply:
The gamerule should be handling what ninja it's attached to and if needed an inverse for the ninja storing its game rule.

Smug is going to take a look too.

Reply:
gamerule references ninja minds and ninja component references the gamerule, probably the right way